std-bits/base/src/main/java/org/enso/base/time/EnsoDateTimeFormatter.java

```diff
@@ -0,0 +1,234 @@
+      return builder.toString();
+    }
+
+    return dateString;
+  }
+
+  @Override
+  public String toString() {
+    return switch (formatterKind) {
+      case SIMPLE -> originalPattern;
+      case ISO_WEEK_DATE -> "(ISO Week Date Format) " + originalPattern;
+      case RAW_JAVA -> "(Java Format Pattern) " + originalPattern;
+      case CONSTANT -> originalPattern;
+    };
+  }
+
+  public LocalDate parseLocalDate(String dateString) {
+    dateString = normaliseInput(dateString);
+
+    TemporalAccessor parsed = formatter.parse(dateString);
+
+    if (parsed.isSupported(ChronoField.EPOCH_DAY)) {
+      return LocalDate.ofEpochDay(parsed.getLong(ChronoField.EPOCH_DAY));
+    }
+
+    // Allow Year and Month to be parsed without a day (use first day of month).
+    if (parsed.isSupported(ChronoField.YEAR) && parsed.isSupported(ChronoField.MONTH_OF_YEAR)) {
+      int dayOfMonth =
+          parsed.isSupported(ChronoField.DAY_OF_MONTH) ? parsed.get(ChronoField.DAY_OF_MONTH) : 1;
+      return LocalDate.of(
+          parsed.get(ChronoField.YEAR), parsed.get(ChronoField.MONTH_OF_YEAR), dayOfMonth);
+    }
+
+    // Allow Year and Quarter to be parsed without a day (use first day of the quarter).
+    if (parsed.isSupported(ChronoField.YEAR) && parsed.isSupported(IsoFields.QUARTER_OF_YEAR)) {
+      int dayOfQuarter =
+          parsed.isSupported(IsoFields.DAY_OF_QUARTER) ? parsed.get(IsoFields.DAY_OF_QUARTER) : 1;
+      int year = parsed.get(ChronoField.YEAR);
+      int quarter = parsed.get(IsoFields.QUARTER_OF_YEAR);
+      int monthsToShift = 3 * (quarter - 1);
+      LocalDate firstDay = LocalDate.of(year, 1, 1);
+      return firstDay.plusMonths(monthsToShift).plusDays(dayOfQuarter - 1);
+    }
+
+    // Allow Month and Day to be parsed without a year (use current year).
+    if (parsed.isSupported(ChronoField.DAY_OF_MONTH)
+        && parsed.isSupported(ChronoField.MONTH_OF_YEAR)) {
+      return LocalDate.of(
+          LocalDate.now().getYear(),
+          parsed.get(ChronoField.MONTH_OF_YEAR),
+          parsed.get(ChronoField.DAY_OF_MONTH));
+    }
+
+    if (parsed.isSupported(IsoFields.WEEK_BASED_YEAR) && parsed.isSupported(IsoFields.WEEK_OF_WEEK_BASED_YEAR)) {
+      // Get the day of week or default to first day if not present.
+      long dayOfWeek = parsed.isSupported(ChronoField.DAY_OF_WEEK) ? parsed.get(ChronoField.DAY_OF_WEEK) : 1;
+      HashMap<TemporalField, Long> fields = new HashMap<>();
+      fields.put(IsoFields.WEEK_BASED_YEAR, parsed.getLong(IsoFields.WEEK_BASED_YEAR));
+      fields.put(IsoFields.WEEK_OF_WEEK_BASED_YEAR, parsed.getLong(IsoFields.WEEK_OF_WEEK_BASED_YEAR));
+      fields.put(ChronoField.DAY_OF_WEEK, dayOfWeek);
+
+      TemporalAccessor resolved = IsoFields.WEEK_OF_WEEK_BASED_YEAR.resolve(fields, parsed, ResolverStyle.SMART);
+      if (resolved.isSupported(ChronoField.EPOCH_DAY)) {
+        return LocalDate.ofEpochDay(resolved.getLong(ChronoField.EPOCH_DAY));
+      }
+    }
+
```

Comment:
I have replicated the logic we had originally for parsing 'partial matches' and even extended it a bit to new cases.

However, for longer term - I'm not sure if this is the right approach. Doing the week-based-year proved to be challenging this way and it seems there is a much simpler way, preferred by the JDK APIs - that is setting defaults for missing fields using `parseDefaulting`. I think ideally we should switch towards this direction - of detecting missing fields at the point of constructing the formatter and adding their defaults there - instead of fighting with the parser here.

I did it for the quarter as there was no other way to do it apparently: 
https://github.com/enso-org/enso/pull/7826/files#diff-9d5efa451e02ef4abbd6920679d8933499c9d53fe9a94cd5e404df8d0ffe8506R50-R54

But I think we could move all of this logic into it, and it would be cleaner.
@jdunkerley what do you think?

I'd like to do it but this PR is huge already so I don't want to introduce any further changes. I'd do it as a followup. It's also low priority since the current solution works OK, it just could be a bit cleaner.

Reply:
Agree - good to do this in the more standard way.

And yes a separate PR would be good!